src/sqlfluff/core/rules/std/L010.py

```diff
@@ -45,99 +47,134 @@ class Rule_L010(BaseCrawler):
+                refuted_cases.update(["pascal"])
+
+        # Update the memory
+        memory["refuted_cases"] = refuted_cases
+
+        self.logger.debug(
+            f"Refuted cases after segment '{segment.raw}': {refuted_cases}"
+        )
+
+        # Skip if no inconsistencies, otherwise compute a concrete policy
+        # to convert to.
+        if cap_policy == "consistent":
+            possible_cases = [c for c in cap_policy_opts if c not in refuted_cases]
+            self.logger.debug(
+                f"Possible cases after segment '{segment.raw}': {possible_cases}"
+            )
+            if possible_cases:
+                # Save the latest possible case and skip
+                memory["latest_possible_case"] = possible_cases[0]
+                self.logger.debug(
+                    f"Consistent capitalization, returning with memory: {memory}"
                 )
-                # Are we just required to be specific?
-                # Policy is either upper, lower or capitalize
-                or (
-                    self.capitalisation_policy != "consistent"
-                    and seen_case != self.capitalisation_policy
+                return LintResult(memory=memory)
+            else:
+                concrete_policy = memory.get("latest_possible_case", "upper")
+                self.logger.debug(
+                    f"Getting concrete policy '{concrete_policy}' from memory"
                 )
-            ):
-                return LintResult(
-                    anchor=segment,
-                    fixes=[
-                        LintFix(
-                            "edit",
-                            segment,
-                            make_replacement(segment, self.capitalisation_policy),
-                        )
-                    ],
-                    memory=memory,
+        else:
+            if cap_policy not in refuted_cases:
+                # Skip
+                self.logger.debug(
+                    f"Consistent capitalization {cap_policy}, returning with "
+                    f"memory: {memory}"
                 )
-            else:
-                # Update memory and carry on
-                cases_seen.add(seen_case)
-                memory["cases_seen"] = cases_seen
                 return LintResult(memory=memory)
+            else:
+                concrete_policy = cap_policy
+                self.logger.debug(
+                    f"Setting concrete policy '{concrete_policy}' from cap_policy"
+                )
 
-        # If it's not a keyword just carry on
-        return LintResult(memory=memory)
+        # We need to change the segment to match the concrete policy
+        if concrete_policy in ["upper", "lower", "capitalise"]:
+            if "pascal" in cap_policy_opts and segment.raw[0].isupper():
+                # Insert undescores in transitions between lower and upper
+                fixed_raw = re.sub("(?<=[a-z0-9])(?=[A-Z])", "_", segment.raw)
```

Comment:
Am I understanding this correctly? Is this adding underscores for pascal case?

Reply:
This segment is the first step in auto-converting PascalCase to one of the other cases. Only PascalCase so far has legal transitions between a lowercase and an uppercase character, so it's always safe to insert an underscore where we see such a transition and the first character of the identifier is uppercase.

This helps the below tests pass:

```
test_pascal_10:
  fail_str: SELECT PasCalCase
  fix_str: SELECT pas_cal_case
  configs:
    rules:
      L014:
        extended_capitalisation_policy: lower

test_pascal_11:
  fail_str: SELECT PascalCaseNAME
  fix_str: SELECT PASCAL_CASE_NAME
  configs:
    rules:
      L014:
        extended_capitalisation_policy: upper
```